torch/csrc/autograd/functions/accumulate_grad.h

```diff
@@ -11,7 +11,83 @@ struct TORCH_API AccumulateGrad : public Node {
 
   variable_list apply(variable_list&& grads) override;
 
+  // Given a variable with its current grad as variable_grad, accumulates
+  // new_grad into variable_grad if in place accumulation is possible.
+  // Otherwise, uses 'update_grad' to update the grad for the variable.
+
+  // variable: the variable whose grad we're accumulating.
+  // variable_grad: the current grad for the variable.
+  // new_grad: new grad we want to acummulate for the variable.
+  // num_expected_refs: the number of refs we expect to hold internally
+  //                    such that it is safe to avoid cloning the grad
+  //                    if use_count() of the grad is less than or equal
+  //                    to this value (in addition to post_hooks).
+  // update_grad: Function that is used to update grad for the variable.
+  //              The argument to the function is a Tensor which
+  //              is used to set a new value for the grad.
+  template <typename T>
+  static void accumulateGradAndCallHooks(
+      const Variable& variable,
+      at::Tensor variable_grad,
+      at::Tensor new_grad,
+      size_t num_expected_refs,
+      const T& update_grad) {
```

Comment:
Do we want to spell `T` out to avoid this function template and move the impl to .cpp file?

```
...
std::function<void(at::Tensor&& grad_update)>&& update_grad) {
```

Reply:
If you spell it as `std::function` we have to do an expensive std::function construction every time you want to call this function. Templating lets the compiler inline the lambda so no construction happens.